### PR TITLE
Add address validation for empty fields

### PR DESCRIPTION
**Pull request by rinnhofer**

### CODE REVIEW

File: Gateway/Helper/ThreeDsHelper.php [hunk 1]
```diff
@@ -141,12 +142,14 @@ private function fetchAccountHolder($address)
         $accountHolder->setEmail($address->getEmail());
         $accountHolder->setPhone($address->getTelephone());
 
-        $sdkAddress = new Address($address->getCountryId(), $address->getCity(), $address->getStreetLine(1));
-        if (!empty($address->getStreetLine(2))) {
-            $sdkAddress->setStreet2($address->getStreetLine(2));
+        if (AddressFactory::isValidAddress($address)) {
```
[the remark this answers is not in the document]
Comment (rinnhofer) [reply Gateway/Helper/ThreeDsHelper.php]: if address invalid - no address should be send

File: Gateway/Validator/QuoteAddressValidator.php [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+     * @return bool
+     * @since 2.2.1
+     */
+    public function validate(array $validationSubject)
+    {
+        $isValid = true;
+
+        /** @var Address $addressObj */
+        $addressObj = $validationSubject['addressObject'];
+        if (empty($addressObj->getCountryId())
```
Comment (ozada) [inline Gateway/Validator/QuoteAddressValidator.php]: You could encapsulate validation for empty of this fields to extra method (isRequiredFieldsNotEmpty for example) in AddressAdapterInterfaceValidator class and extend QuoteAddressValidator from it.

File: Gateway/Helper/ThreeDsHelper.php [hunk 3]
```diff
@@ -40,17 +41,25 @@ class ThreeDsHelper
...
      * @param AccountHolderFactory $accountHolderFactory
+     * @param QuoteAddressValidator $addressValidator
+     *
+     * @since 2.2.1 added QuoteAddressValidator
      */
     public function __construct(
         AccountInfoFactory $accountInfoFactory,
-        AccountHolderFactory $accountHolderFactory
+        AccountHolderFactory $accountHolderFactory,
+        QuoteAddressValidator $addressValidator
```
[the remark this answers is not in the document]
Comment (rinnhofer) [reply Gateway/Helper/ThreeDsHelper.php]: added factory to handle the various validators

File: Gateway/Validator/QuoteAddressValidator.php [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+     * @param array $validationSubject
+     * @return bool
+     * @since 2.2.1
+     */
+    public function validate(array $validationSubject)
+    {
+        $isValid = true;
+
+        /** @var Address $addressObj */
+        $addressObj = $validationSubject['addressObject'];
```
[the remark this answers is not in the document]
Comment (rinnhofer) [reply Gateway/Validator/QuoteAddressValidator.php]: added constructor with usage of factory first

File: Gateway/Validator/AddressAdapterInterfaceValidator.php [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+     * @param array $validationSubject
+     * @return bool
+     * @since 2.2.1
+     */
+    public function validate(array $validationSubject)
+    {
+        $isValid = true;
+
+        /** @var AddressAdapterInterface $addressObj */
+        $addressObj = $validationSubject['addressObject'];
```
[the remark this answers is not in the document]
Comment (rinnhofer) [reply Gateway/Validator/AddressAdapterInterfaceValidator.php]: skip abstract address validator due to hopefully removal of additional address object within the next release

File: Test/Unit/Gateway/Validator/AddressAdapterInterfaceValidatorUTest.php [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    public function testValidAddressObject()
+    {
+        $this->magentoAddressInterface->method('getCountryId')->willReturn('AT');
+        $this->magentoAddressInterface->method('getCity')->willReturn('Testcity');
+        $this->magentoAddressInterface->method('getStreetLine1')->willReturn('Teststreet 1');
+
+        $actual = new AddressAdapterInterfaceValidator($this->magentoAddressInterface);
+        $this->assertTrue($actual->validate([]));
```
Comment (ozada) [inline Test/Unit/Gateway/Validator/AddressAdapterInterfaceValidatorUTest.php]: delete [] from -> validate.

**Pull request merged**